README.rst

```diff
@@ -1,103 +1,262 @@
-stream.
+This module implements a DSL for creating binary parsers/encoders.
+It exports the macro ``createParser`` which generates a ``tuple[get: proc, put: proc]``.
+``get`` returns a tuple with each parsed field and ``put`` writes a compatible tuple to the stream.
+
+The macro accepts 3 kind of things:
+
+-  Parser options
+-  Parser parameters
+-  Block with DSL statements
+
+Parser options
+--------------
+
+Each specified option must be in the form ``option = value``:
+
+- ``endian``: sets the default endianness for the whole parser
+   - *default*: big endian
+   - ``b``: **big** endian
+   - ``l``: **little** endian
+- ``bitEndian``: sets the default bit-endianness for the whole parser
+   - *default*: left -> right
+   - ``n``: left -> right (**normal**)
+   - ``r``: left <- right (**reverse**)
+
+Parser parameters
+-----------------
+
+Each parameter must be in the form ``symbol: type``. The generated
+``get``/``put`` procs will then have this additional parameter appended.
+
+DSL
+----
+
+Each statement corresponds to 1 field. The general syntax is:
+
+.. code:: nim
+
+    Type {Operations}: Value
+
+where ``{Operations}`` is optional and refers to a plugin system (see
+below).
+
+Type
+~~~~
+
+The **kind**, **endianness** and **size** are encoded in a identifier
+made up of:
+
+- 1 optional letter specifying the kind:
+   - *default*: signed integer
+   - ``u``: **unsigned** integer
+   - ``f``: **float**
+   - ``s``:**string**
+   - ``*``: complex (see below)
+- 1 optional letter specifying endianness:
+   - *default*: big endian
+   - ``b``: **big** endian
+   - ``l``: **little** endian
+- 1 optional letter specifying big-endianness for unaligned reads:
+   - *default*: left -> right
+   - ``n``: left -> right (**normal**)
+   - ``r``: left <- right (**reverse**)
+- 1 number specifying size in **bits**:
+   - for strings only byte multiples are allows (``8``, ``16``, ``24``, ...)
+   - for an integer the allowed values are ``1 .. 64``
+   - for a float the allowed values are ``32`` and ``64``
+   - for a custom it can't be used (use the secondary ``size`` operation)
+
+| Strings may not have a size, in which case they are null-terminated.
+| You can order options however you want, but size must come last (e.g. ``lru16`` and ``url16`` are valid but not ``16lru``).
+
+Value
+~~~~~
+
+This section consists of a mandatory name for the field + optional
+repetition + optional assertion.
+
+For primitive types (except null-terminated strings) you can instruct
+binaryparse to not produce a symbol and discard the field by using ``_``
+for the name:
+
+.. code:: nim
+
+    createParser(magic):
+      s: _ = "GIF89a"
+
+Alignment
+~~~~~~~~~
+
+Currently unaligned reads for strings are not supported:
+
+.. code:: nim
+
+    createParser(myParser):
+      6: x
+      s: y # invalid, generates an exception
+
+| Endian only refers to aligned reads, while bit-endian only refers to unaligned reads.
```

Comment:
This should probably be fixed at some point. There's no reason why endian shouldn't work on non-aligned fields.

Reply:
Fixed.